cumulus/tasks/sf-sns-report/index.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+const get = require('lodash.get');
+const { StepFunction } = require('@cumulus/ingest/aws');
+const { setGranuleStatus, sns } = require('@cumulus/common/aws');
+const errors = require('@cumulus/common/errors');
+const cumulusMessageAdapter = require('@cumulus/cumulus-message-adapter-js');
+
+/**
+ * Determines if there was a valid exception in the input message
+ *
+ * @param {Object} event - aws event object
+ * @returns {boolean} true if there was an exception, false otherwise
+ */
+function eventFailed(event) {
+  // event has exception
+  // and it is needed to avoid flagging cases like "exception: {}" or "exception: 'none'"
+  if (event.exception && (typeof event.exception === 'object') &&
+    (Object.keys(event.exception).length > 0)) {
+    return true;
+  }
+  // Error and error keys are not part of the cumulus message
+  // and if they appear in the message something is seriously wrong
+  else if (event.Error || event.error) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * if the cumulus message shows that a previous step failed,
+ * this function extracts the error message from the cumulus message
+ * and fails the function with that information. This ensures that the
+ * Step Function workflow fails with the correct error info
+ *
+ * @param {Object} event - aws event object
+ * @returns {undefined} throws an error and does not return anything
+ */
+function makeLambdaFunctionFail(event) {
```

Comment:
It's a little hard to follow what this function is doing.  It might be easier with something like:

```
function buildError(type, cause) {
  let ErrorClass;

  if (Object.keys(errors).includes(type)) ErrorClass = errors[type];
  else if (type === 'TypeError') ErrorClass = TypeError;
  else ErrorClass = Error;

  return new ErrorClass(cause);
}

function makeLambdaFunctionFail(event) {
  const error = event.exception || event.error;

  if (!error) throw new Error('Step Function failed for an unknown reason.');
  
  throw buildError(error.Error, error.Cause);
}
```

Reply:
Thanks.  This makes the code much clear.  Sorry I didn't try to understand the code.